Add option to advect enthalpy instead of cp T for non constant cp fluids

GiudGiud opened the pull request:
We already had a correction in the time derivative for non-constant cp. But @freiler spotted that we did not in the advection term.

This should eventually be the default. I ll do that when I transition the nonlinear fluid heat transfer Physics to use this formulation

GiudGiud committed:
Add an option to use a FP UO to compute the enthalpy
refs #12174

GiudGiud committed:
Add a test for non-constant cp

tanoret commented:
Please add realistic testing or more extended verification against analytical solutions. Let's try avoiding the same issues that we had in the past. Thank you!

tanoret commented:
Please add the same tests than for the linear problem:

- [ ] 1D problem with cp=a+bT and verification against analytical solution
- [ ] 2D problem with wall heating and verifying enthalpy balances (h_dot_out = h_dot_in + q''_wall * A_wall)
- [ ] Equivalent problem in 3D and test against the previous 2D problem (this is important to see that we are not blowing up the size of the AD container by advecting enthalpy)
- [ ] Body with variable cp and rho cooling at a fixed heat rate and verify against the analytical solution (this is verifying the implementation of the time derivative)

Thank you!

GiudGiud commented:
we should be good for this one

> Equivalent problem in 3D and test against the previous 2D problem (this is important to see that we are not blowing up the size of the AD container by advecting enthalpy)

whether rho cp T or h, you have the same dependencies on pressure and temperature. We are not adding new ones by switching to h

For the time derivative we probably want to do more work than this if we want a conservative scheme. 
I ll see what I can do for 1D / 2D

freiler commented:
We can use the 1D and 2D cases I have for enthalpy to eventually check the accuracy of the new kernel. In addition, I will also build a transient case to test the time derivative kernel.

GiudGiud commented:
@freiler do you want to test this PR to provide 1 and 2? Or should I use your inputs?

freiler commented:
I can push the input for the 1D  and 2D cases to your branch so we at least run the same test cases and compare.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2607707/), step Generate coverage on c488692 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage generation failed unexpectedly

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>4e4e5d</th>
      <th colspan=4>#29569 c48869</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.72%</td>
      <td>84.71%</td>
      <td>-0.01%</td>
      <td>82.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17494</td>
      <td>17515</td>
      <td>+21</td>
      <td>33</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3156</td>
      <td>3162</td>
      <td>+6</td>
      <td>7</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29569/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29569/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29569/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 82.50% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2607692/), step Docs: sync website on c488692 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29569/site)

This comment will be updated on new commits.

GiudGiud committed:
Add an option to input h directly so we can use user-input FPs

GiudGiud committed:
Add cp derivatives to NaCl, was done for testing purposes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2609894/), step Clang format on f21f9ce wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dc0e9c8610932f7d6b4919833db28a710b949790`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2609894/), step Fixup commit check on f21f9ce wanted to post the following:

The following commit(s) contain fixups:

- 69b87c6b36: fixup! Add an option to use a FP UO to compute the enthalpy refs #12174

GiudGiud committed:
Add 1D case showing:
- balance to 1e-8
- error to T_out in analytical solution down to 5e-6

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2610319/), step Clang format on 0a0aa7d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dc0e9c8610932f7d6b4919833db28a710b949790`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2610319/), step Fixup commit check on 0a0aa7d wanted to post the following:

The following commit(s) contain fixups:

- 69b87c6b36: fixup! Add an option to use a FP UO to compute the enthalpy refs #12174

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2610838/), step Clang format on d0d9f2f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a9911ed8fb32d71d4dcbe970680b2f1c8a8fa2c2`

GiudGiud committed:
Add a 2D case showing the energy balance in 2D with wall heating

GiudGiud commented:
I think the transient case should be done in a PR on time integration. As explained in the thread on slack, I dont expect the time derivative discretization rho_new cp_new T_new - rho_new cp_new T_old to be conservative.

lindsayad commented:
@freiler @tanoret can one of you review this please?

grmnptr commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -32,31 +33,106 @@ INSFVEnthalpyFunctorMaterial::validParams()
-  addFunctorProperty<ADReal>(
-      "rho_cp_temp", [&rho_h](const auto & r, const auto & t) -> ADReal { return rho_h(r, t); });
+    addFunctorProperty<ADReal>(
+        "rho_cp_temp", [&rho_h](const auto & r, const auto & t) -> ADReal { return rho_h(r, t); });
```
If we are making this more general, I'd not call it `rho_cp_temp` anymore. Most of the code fetches this under the hood so I hope not a lot of input files need changing.

GiudGiud replied:
ok let's just not define these functors anymore. 

added the transient test too

grmnptr reviewed (commented):
Looks good to me! - did the code review. I think the test cases are also convincing. 

`INSFVEnergyTimeDerivative` should be correct if you pass the new functor to it. 
`WCNSFVEnergyTimeDerivative` should also work I think.

Could you please add a test to verify that they work?

GiudGiud commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -23,7 +24,7 @@ INSFVEnthalpyFunctorMaterial::validParams()
       "This is the material class used to compute enthalpy for the "
       "incompressible/weakly-compressible finite-volume implementation of the Navier-Stokes "
-      "equations. Note that this class assumes that cp is a constant");
+      "equations.");
```
I havent been keeping track with his PR but I think his functor material might be focusing on computing T from h rather than h from T

tanoret replied:
If the class assumes that cp is constant, please add it to the name of the functormaterial, e.g., INSFVConstantCpEnthalpyFunctorMaterial or something. Otherwise, this is can lead to errors

GiudGiud replied:
the assumptions depend on the parameter

grmnptr replied:
I checked and I think the main difference is that this works with AD, the one in the other implementation is nonAD. So I am fine having 2 materials.

GiudGiud committed:
Address Peter's review:
- stop defining cpT and rho cp T functors
- add a transient test
+ add an exception test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2632983/), step Clang format on dfcac83 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29569/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e193a3d452b52030c368f8b0e4cb692ef0113925`

tanoret commented on modules/fluid_properties/src/fluidproperties/NaClFluidProperties.C:
```diff
@@ -137,19 +137,42 @@ NaClFluidProperties::cp_from_p_T(Real pressure, Real temperature) const
+    Real pressure, Real temperature, Real & cp, Real & dcp_dp, Real & dcp_dT) const
+{
+  // Correlation needs pressure in bar
+  Real pbar = pressure * 10.0e-5;
```
Conversion to bar should be 1e-5 (1 bar = 1e5 Pa)

GiudGiud replied:
very true. this is like this all over this property :O

GiudGiud replied:
fixed

tanoret commented on modules/fluid_properties/src/fluidproperties/NaClFluidProperties.C:
```diff
@@ -137,19 +137,42 @@ NaClFluidProperties::cp_from_p_T(Real pressure, Real temperature) const
+  // Halite isobaric heat capacity
+  cp = 1148.81 + 0.551548 * (Tc - Tt) + 2.64309e-4 * (Tc - Tt) * (Tc - Tt) + r3 * pbar +
+       r4 * pbar * pbar;
+  dcp_dp = r3 * 10.e-5 + 2 * r4 * pbar * 10.e-5;
```
Conversion to bar may also be wrong here

GiudGiud replied:
fixed

tanoret commented on modules/fluid_properties/src/fluidproperties/NaClFluidProperties.C:
```diff
@@ -137,19 +137,42 @@ NaClFluidProperties::cp_from_p_T(Real pressure, Real temperature) const
 
+void
+NaClFluidProperties::cp_from_p_T(
+    Real pressure, Real temperature, Real & cp, Real & dcp_dp, Real & dcp_dT) const
```
Could we please add unit test to compare the computed cp against tabulated values for a few points in the input sapce?

GiudGiud replied:
they already exist, the dependence on pressure is just so small it did not catch it

tanoret commented on modules/navier_stokes/include/functormaterials/INSFVEnthalpyFunctorMaterial.h:
```diff
@@ -23,12 +25,24 @@ class INSFVEnthalpyFunctorMaterial : public FunctorMaterial
 
 protected:
+  /// whether we can use a constant cp as a shortcut to compute enthalpy
+  bool _assume_constant_cp;
```
```suggestion
  bool _assumed_constant_cp;
```

tanoret commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -32,31 +33,92 @@ INSFVEnthalpyFunctorMaterial::validParams()
       NS::specific_enthalpy, NS::specific_enthalpy, "the name of the specific enthalpy");
+
+  // To handle non constant cp
+  params.addParam<bool>("assume_constant_cp", true, "Whether to assume cp is constant");
```
```suggestion
  params.addParam<bool>("assumed_constant_cp", true, "Whether to assume cp is constant");
```

tanoret commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -32,31 +33,92 @@ INSFVEnthalpyFunctorMaterial::validParams()
 
 INSFVEnthalpyFunctorMaterial::INSFVEnthalpyFunctorMaterial(const InputParameters & parameters)
   : FunctorMaterial(parameters),
+    _assume_constant_cp(getParam<bool>("assume_constant_cp")),
```
```suggestion
    _assumed_constant_cp(getParam<bool>("assumed_constant_cp")),
```

tanoret commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -32,31 +33,92 @@ INSFVEnthalpyFunctorMaterial::validParams()
-                                 [this](const auto & r, const auto & t)
-                                 { return _rho(r, t) * _cp(r, t) * _temperature(r, t); });
+  // We have to use a warning because fp is often in the global parameters
+  if (_assume_constant_cp && _fp)
```
```suggestion
  if (_assumed_constant_cp && _fp)
```

tanoret commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -32,31 +33,92 @@ INSFVEnthalpyFunctorMaterial::validParams()
+               "Must specify both fluid properties and pressure or an enthalpy functor if not "
+               "assuming the specific enthalpy is constant");
+
+  if (_assume_constant_cp)
```
```suggestion
  if (_assumed_constant_cp)
```

tanoret commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/1d_test_cpT.i:
```diff
@@ -0,0 +1,231 @@
+
+    assume_constant_cp = false
+    h_in = 'h'
+    # fp = 'fp'
+    # pressure = 'pressure'
```
Remove is unused

GiudGiud replied:
it's the other way of setting this up. I ll add a comment

tanoret commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/1d_test_cpT.i:
```diff
@@ -0,0 +1,231 @@
+    variable = T
+    boundary = 'right'
+  []
+  [T_analytical]
```
```suggestion
  [T_analytical_outlet]
```

tanoret commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/1d_test_cpT.i:
```diff
@@ -0,0 +1,231 @@
+  []
+  [error_T]
+    type = ParsedPostprocessor
+    expression = 'T_out - T_analytical'
+    pp_names = 'T_out T_analytical'
```
```suggestion
    expression = 'T_out - T_analytical_outlet'
    pp_names = 'T_out T_analytical_outlet'
```

tanoret commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/1d_test_cpT.i:
```diff
@@ -0,0 +1,231 @@
+  []
+  [T_analytical]
+    type = Receiver
+    default = ${fparse (-A_cp+sqrt(A_cp^2-2*B_cp*(-q_source/rho/bulk_u*L-A_cp*T_in-B_cp/2*T_in*T_in)))/B_cp}
```
I verified by hand calculation that this analytical solution is correct. However, please add a couple comments of the steps of the derivation for future reference.

GiudGiud replied:
Ramiro did the derivation. @freiler do you want to add that as a suggestion?

tanoret commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/1d_test_cpT.i:
```diff
@@ -0,0 +1,231 @@
+L = 30
```
This is an excellent test! Exactly the kind of things we need. Thank you!

tanoret commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/enthalpy_computation.i:
```diff
@@ -0,0 +1,101 @@
+[Mesh]
```
What is the purpose of this test? Let's compute the enthalpies by hand and compare the code result against the hand calculation. That way we know that things are being done properly

GiudGiud replied:
it's the original regression test. I ll add the hand calc

GiudGiud replied:
done. Had to change to PbBiFPs because Water97 is quite complex in terms of computing a specific fluid property

tanoret commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/2d-steady-wall-balance.i:
```diff
@@ -0,0 +1,261 @@
+    function = ${q2_wall}
+    boundary = 'top'
+  []
+  [balance_in_percent]
+    type = ParsedPostprocessor
+    expression = '(H_out + H_in - Q - Q_wall) / H_in * 100'
+    pp_names = 'H_in H_out Q Q_wall'
```
This test is currently formulated as a non-regression test. However, this is the only postprocessor that we actually need to check. Let's please compare in the test this pp against 0 with a certain tolerance

tanoret reviewed (changes requested):
Thank you for the work. I am suggesting a couple of changes in the tests. Please let me know if further clarification is needed for anything. Thanks again!

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/1d_test_cpT.i:
```diff
@@ -0,0 +1,231 @@
+
+    assume_constant_cp = false
+    h_in = 'h'
+    # fp = 'fp'
```
```suggestion
    # Alternative to providing 'h': set the fluid property and the pressure parameter
    # fp = 'fp'
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/1d_test_cpT.i:
```diff
@@ -0,0 +1,231 @@
+    rho = 'rho'
+    cp = 'cp'
+
+    assume_constant_cp = false
```
```suggestion
    assumed_constant_cp = false
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/tests:
```diff
@@ -39,6 +52,38 @@
+    cli_args = "Outputs/file_base=enthalpy_computation_non_constant_cp_out
+                FunctorMaterials/compute_cp/fp=fp
+                FunctorMaterials/compute_cp/pressure=pressure
+                FunctorMaterials/compute_cp/assume_constant_cp=false"
```
```suggestion
                FunctorMaterials/compute_cp/assumed_constant_cp=false"
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/tests:
```diff
@@ -49,5 +94,12 @@
+    [missing_fp]
+      type = RunException
+      input = '2d-transient.i'
+      cli_args = "FunctorMaterials/ins_fv/assume_constant_cp=false"
```
```suggestion
      cli_args = "FunctorMaterials/ins_fv/assumed_constant_cp=false"
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/wcns/materials/gold/1d_test_cpT_out.csv:
```diff
@@ -0,0 +1,3 @@
+time,H_in,H_out,Q,T_analytical,T_out,balance_in_percent,error_T
```
```suggestion
time,H_in,H_out,Q,T_analytical_outlet,T_out,balance_in_percent,error_T
```

GiudGiud committed:
Address Mauricio's review

- rename a boolean
- rename a postprocessor
- fix conversion to bar in NaCl cp computation
- simplify enthalpy computation test to add a "hand" calc of the enthalpy

Co-authored-by: Mauricio Tano <mauriciotano@gmail.com>

grmnptr commented on modules/navier_stokes/src/functormaterials/INSFVEnthalpyFunctorMaterial.C:
```diff
@@ -32,31 +33,92 @@ INSFVEnthalpyFunctorMaterial::validParams()
+    addFunctorProperty<ADReal>(
+        "rho_cp_temp", [&rho_h](const auto & r, const auto & t) -> ADReal { return rho_h(r, t); });
 
-  const auto & h = addFunctorProperty<ADReal>(NS::specific_enthalpy,
-                                              [this](const auto & r, const auto & t)
-                                              { return _cp(r, t) * _temperature(r, t); });
+    addFunctorProperty<ADReal>("cp_temp",
+                               [&h](const auto & r, const auto & t) -> ADReal { return h(r, t); });
```
Did these stay because you don't want to modify inputs/other source files that grab these under the hood? I think it would be nicer if we didn't have these and had something uniform.

grmnptr replied:
Sounds good!

grmnptr reviewed (commented):
Great! Just one more question:

GiudGiud commented:
Thanks for the review!
